### PR TITLE
types(python): Feature Flag types

### PR DESCRIPTION
**Pull request by mgaeta**
This PR adds type annotations to the `src/sentry/features/` directory. Unfortunately I ran into the same circular dependency problem with src/sentry/models so I couldn't user `User`, `Project`, and `Organization`. There is a second circular dependency with `handler.py` -> `manager.py` that prevents annotating with the `FeatureHandler ` and `FeatureManager` types.

### CODE REVIEW

File: src/sentry/features/base.py [hunk 1]
```diff
@@ -1,24 +1,26 @@
 __all__ = ["Feature", "OrganizationFeature", "ProjectFeature", "ProjectPluginFeature"]
 
+from typing import Any
+
 
 class Feature:
-    def __init__(self, name):
+    def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
```
Comment (evanh) [inline src/sentry/features/base.py:7]: Why does this need `*args` and `**kwargs`?

File: src/sentry/features/handler.py [hunk 2]
```diff
@@ -28,13 +43,13 @@ def has_for_batch(self, batch):
...
 
 class BatchFeatureHandler(FeatureHandler):
-    def _check_for_batch(self, feature_name, organization, actor):
+    def _check_for_batch(self, feature_name: str, organization: Any, actor: Any) -> bool:
         raise NotImplementedError
 
-    def has(self, feature, actor):
-        organization = getattr(feature, "organization", None) or feature.project.organization
+    def has(self, feature: Union[Feature], actor: Any) -> bool:
+        organization = getattr(feature, "organization", None) or feature.project.organization  # type: ignore
```
Comment (evanh) [inline src/sentry/features/handler.py:50]: Why the ignore here?

File: src/sentry/features/helpers.py [hunk 3]
```diff
@@ -1,18 +1,24 @@
 from typing import Any, Sequence
 
+from rest_framework.request import Request
 from rest_framework.response import Response
 
 from sentry import features
 from sentry.models import Organization
 
+# TODO(mgaeta): It's not currently possible to make a Callable with kwargs.
```
Comment (evanh) [inline src/sentry/features/helpers.py:9]: See https://mypy.readthedocs.io/en/stable/kinds_of_types.html#callable-types-and-lambdas.

I'm not sure what exactly you're trying to do here, but something like `Callable[..., T]` is possible if you want to specify a function that takes arbitrary arguments and returns T.

**Review by evanh — COMMENTED** [review]
About the import cycle, mypy has a solution for that: https://mypy.readthedocs.io/en/latest/runtime_troubles.html#import-cycles

You could use that, as well as annotating the functions with strings (e.g. `-> "Actor"` instead of `-> Actor`) to solve that problem.

**Comment by mgaeta** [comment]
`TYPE_CHECKING` looks really interesting I'll give it a try

File: src/sentry/features/handler.py (hunk 2 above)
Comment (mgaeta) [reply src/sentry/features/handler.py]: I was trying to get around the `Feature` not having `.project` but I solved this with `.get_organization()`.

File: src/sentry/features/base.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/features/base.py:11]: I left a note in the constructor's docstring. In `FeatureManager.get()` we do:
```
cls = self._get_feature_class(name)
return cls(name, *args, **kwargs)
```
to handle passing in `organization`, `project`, and `plugin`.

File: src/sentry/features/helpers.py [hunk 4]
```diff
@@ -1,18 +1,24 @@
-from typing import Any, Sequence
+from typing import Any, Callable, Optional, Sequence
 
+from rest_framework.request import Request
 from rest_framework.response import Response
 
 from sentry import features
 from sentry.models import Organization
 
+# TODO(mgaeta): It's not currently possible to type a Callable's args with kwargs.
```
Comment (evanh) [inline src/sentry/features/helpers.py:9]: 💀

**Review by evanh — APPROVED** [review]
Looks good to me.

**Pull request merged**